find_email_addresses.py

```diff
@@ -0,0 +1,156 @@
+    found or an error occurs -1 will be returned.  If a timeout error occurs,
+    this will re-attempt opening the page 3 times before giving up.
+    """
+
+    attempts = 0
+
+    while attempts < 3:
+        try:
+            resp = urlopen(url)
+            return resp
+        except URLError, e:
+            # If we get a 408, try again, on any other error exit
+            if hasattr(e, 'code') and e.code == 408:
+                attempts += 1
+            else:
+                attempts = 3
+
+    return -1
+
+
+def getEmails(s):
+    """
+    Find and return all email addresses in the given string s.
+    """
+
+    # A regex pattern for email addresses
+    email_pattern = re.compile(r"([A-Za-z0-9#\-_~!$&'()*+,;=:]+"
+                               "(?:\.[A-Za-z0-9#\-_~!$&'()*+,;=:]+)*"
+                               "@[A-Za-z0-9\-]+\.[A-Za-z0-9\-]+)")
+    emails = email_pattern.findall(s)
+
+    # Strip off the "mailto:" before returning, for cases where an email link
+    # is given, but not the email text
+    emails = [email[7:] if email.startswith("mailto:") else email
+              for email in emails]
+    return emails
+
+
+def getLinks(s, domain):
+    """
+    Find and return all clickable links in the given string s.
+    """
+
+    # Find all links of the form <a href="(link)"
+    link_pattern = re.compile(r"""<a[^<>]+href\s*=\s*"([^"]+)""")
+    links = list(set(link_pattern.findall(s)))
+
+    # Add the domain to relative links
+    links = [parse.urljoin(domain, link) for link in links]
+
+    # Finally, remove any links not in the domain or image links
+    links = [link for link in links
+             if parse.urlparse(link).geturl().startswith(domain) and
+             not (link[-4:] in [".png", ".jpg", ".gif"])]
+
+    return links
+
+
+def crawlForEmail(url, domain="", visited=set()):
+    """
+    Recursive function that scans for emails in the text on the page at url,
+    then recursively calls on each link on that page in the same domain that
+    is not in the visited set, and returns all emails found in this way.
+    """
+
+    # First get all text, and get all emails and links
+    resp = getPage(url)
+    if resp == -1:
+        # On error, return the empty set; no emails found
+        return set()
+
+    rtext = resp.read()
+    if domain == "":
+        # If no domain was given, the url was the domain
+        domain = resp.geturl()
+        visited.add(domain)
+
+    # Find the emails and links from the page text
+    emails = set(getEmails(rtext))
+    links = getLinks(rtext, domain)
+
+    # Remove any links already visited
+    links = [link for link in links if link not in visited]
+
+    # Add all of the links intended to visit to the visited set
+    visited.update(set(links))
+
+    for link in links:
+        # Recur on this function, adding any found emails to the email set
+        emails.update(crawlForEmail(link, domain, visited))
```

Comment:
On the topic of microsoft.com, the reason it wasn't returning anything is because I didn't specify a User-Agent so the server returned very basic output and there was nothing to crawl. This can be fixed by specifying a User-Agent and url in a Request object and passing that into urllib.urlopen instead, which gives full access to Microsoft's site.

Unfortunately, even with that change it doesn't fully work.  Something about following the drop down menu items continually generates an infinite number of drop down link variants to follow, and I wasn't able to narrow down exactly what the problem is to make it scrape properly.